### DISCUSSION TITLE
Plans for query time performance improvements

### DISCUSSION THREAD

**Discussion by rrva** [graphql-java/graphql-java#3220] [General]
We have an app which resolves most of the fields from memory, and we consistently see graphql-java engine code show up when profiling in production. Grapqhl-java itself is the bottleneck for our throughput. In #3175 we started considering implementing our own execution strategy. Is there any performance work planned for query execution? Virtual threads is planned to reach GA end of September, so if virtual threads is one ingredient for improving execution, the timing is great to start looking at that now. Any other ideas around query execution performance?

**Comment by bbakerman** [comment]
>  Is there any performance work planned for query execution?

Not specifically - the last 2 years have been a series of performance work but we don't have plans for a major rewrite.

> Virtual threads is planned to reach GA end of September, so if virtual threads is one ingredient for improving execution, the timing is great to start looking at that now.

Its not known if virtual threads will make a significant difference.  They are more about having a cleaner programming model than being faster.  It maybe that they are faster but I cant imagine it will  be significantly faster.  But let's see.

We have had a rought spike on Loom a year or so ago - https://github.com/graphql-java/graphql-java/pull/2831/

It showed its possible but requires a separate engine if we are to retain Java 11 support say for those not on the latest.

That spike was not performance tested however that I know of - but rather what code shape would work

So far our spitballing on this is to create a new engine that compiles ONLY on say Java 21 - this would involve the graphql-java becoming a multi artifact project (it currently isnt but it could be) so we we would have some of the code being in Java 11 and some other in a future version.

The other area what would need to change with a seperate engine is `Instrumentation` since it to is based on `CompleteableFuture` as well.  So a new engine would need a matching new `Instrumentation`
